### ISSUE TITLE
flaskapp redeployement

### ISSUE THREAD

**Issue by al3xfulton** [picoCTF/picoCTF#254]
When creating a flaskapp problem changes, any changes after the initial deployment lag in their response to the user (if at all). It appears that even though the underlying code changes in the directory it doesn't stop/start the application until forced to by another trigger.

**Comment by shombo** [comment]
>  lag in their response to the user

Can you elaborate on this? Do you mean that the web server performance is degraded (HTTP requests take a long time to be fulfilled), or do you mean that the changes are not reflected at all until a Flask / webserver restart?

In either case, is this in reference to what should be an already deployed / developed challenge?

Lastly, does this [built in functionality](http://flask.pocoo.org/docs/1.0/quickstart/#debug-mode) solve this issue?

**Comment by syreal17** [comment]
@shombo: @al3xfulton has graduated, but I can try to elaborate based on my experience. @dmartin may know more.

OP is indicating as you suggested that no changes to problems are visible to participants until something else causes the web service to restart. As I related in an email, @dmartin recently did a huge refactor on the web server code, though I wouldn't guess that this issue would be affected. I *would indeed* guess that enabling debugging mode as suggested by the link you posted would fix this issue, however such functionality is quite the stepping stone to remote code execution as also warned in the linked documentation. I wouldn't want to run flask in debug mode in production. I'm guessing that this is relation to operational concerns (modifying a problem during a competition) and perhaps mostly could be fixed by an automatic restart or at least a warning about no changes being actuated until the next restart.

So what's your concern with this particular issue?

**Comment by shombo** [comment]
I guess I was trying to determine whether or not this was related more to development or production. 

I would never advise to run flask in debug mode in production here. But if the issue was regarding developing a challenge / debugging its deployment in a locally "deployed" environment, then flask has a built in solution. Having a debug mode may be useful for this case, which could just flask in debug mode. 

What you're describing seems to be related more to needing to patch a currently deployed / live challenge. This would indeed be an issue requiring a restart of the web server (nginx). 

Lastly, this popped onto my radar because I was just looking for issues to potentially tackle.

**Comment by syreal17** [comment]
Yes, I'm largely guessing on most on my thoughts on this. It's probably an issue for both development and production, but it's great to be aware of debug mode for flask, in development. Remind me on Discord and I'll reach out to Alex if you're considering this issue after seeing some of the others.

The project does have a lot of churn as we employ students and interns and also it has gone from a PhD dissertation to what it is now!

**Comment by al3xfulton** [comment]
The initial intent of this was almost exclusively on the development side, while iterating through minor changes in the code it would often not reflect on the server side, which caused some real headaches. While the process that we used for 2018 (at least on the problems I developed) certainly started with using stand alone flask in debug mode, it was never perfect, and finding a typo or making a change during alpha/beta testing became frustrating as it required (relatively) significant restarting. 

There was a bit of a disconnect between what appeared using straight flask/debug mode  and what actually spawned using the picoCTF framework. This is may have been a separate problem from what the initial post was about, but debugging the uwsgi deployment caused some serious headache as well trying to make that transition to the framework as that really didn't log anywhere unless the uwsgi command was extracted from the batch file that was actually launched (with some modifications for uid and gid that can cause a good bit of difference).
